--- Day 13: Claw Contraption ---
Next up: the lobby of a resort on a tropical island. The Historians take a moment to admire the hexagonal floor tiles before spreading out.

Fortunately, it looks like the resort has a new arcade! Maybe you can win some prizes from the claw machines?

The claw machines here are a little unusual. Instead of a joystick or directional buttons to control the claw, these machines have two buttons labeled A and B. Worse, you can't just put in a token and play; it costs 3 tokens to push the A button and 1 token to push the B button.

With a little experimentation, you figure out that each machine's buttons are configured to move the claw a specific amount to the right (along the X axis) and a specific amount forward (along the Y axis) each time that button is pressed.

Each machine contains one prize; to win the prize, the claw must be positioned exactly above the prize on both the X and Y axes.

You wonder: what is the smallest number of tokens you would have to spend to win as many prizes as possible? You assemble a list of every machine's button behavior and prize location (your puzzle input). For example:

Button A: X+94, Y+34
Button B: X+22, Y+67
Prize: X=8400, Y=5400

Button A: X+26, Y+66
Button B: X+67, Y+21
Prize: X=12748, Y=12176

Button A: X+17, Y+86
Button B: X+84, Y+37
Prize: X=7870, Y=6450

Button A: X+69, Y+23
Button B: X+27, Y+71
Prize: X=18641, Y=10279
This list describes the button configuration and prize location of four different claw machines.

For now, consider just the first claw machine in the list:

Pushing the machine's A button would move the claw 94 units along the X axis and 34 units along the Y axis.
Pushing the B button would move the claw 22 units along the X axis and 67 units along the Y axis.
The prize is located at X=8400, Y=5400; this means that from the claw's initial position, it would need to move exactly 8400 units along the X axis and exactly 5400 units along the Y axis to be perfectly aligned with the prize in this machine.
The cheapest way to win the prize is by pushing the A button 80 times and the B button 40 times. This would line up the claw along the X axis (because 80*94 + 40*22 = 8400) and along the Y axis (because 80*34 + 40*67 = 5400). Doing this would cost 80*3 tokens for the A presses and 40*1 for the B presses, a total of 280 tokens.

For the second and fourth claw machines, there is no combination of A and B presses that will ever win a prize.

For the third claw machine, the cheapest way to win the prize is by pushing the A button 38 times and the B button 86 times. Doing this would cost a total of 200 tokens.

So, the most prizes you could possibly win is two; the minimum tokens you would have to spend to win all (two) prizes is 480.

You estimate that each button would need to be pressed no more than 100 times to win a prize. How else would someone be expected to play?

Figure out how to win as many prizes as possible. What is the fewest tokens you would have to spend to win all possible prizes?

In [ ]:
import re
with open('input13.txt') as file:
    data = []
    for line in file: # all inputs are positive
        if line.startswith('Button A'):
            btn_a = tuple(int(x) for x in re.findall(r'\d+', line)) 
        elif line.startswith('Button B'):
            btn_b = tuple(int(x) for x in re.findall(r'\d+', line))
        elif line.startswith('Prize'):
            prize = tuple(int(x) for x in re.findall(r'\d+', line))
            data.append((btn_a, btn_b, prize))
data

[((40, 38), (21, 84), (4245, 5634)),
 ((19, 11), (37, 98), (3246, 6474)),
 ((12, 38), (44, 28), (812, 12604)),
 ((29, 13), (39, 72), (2243, 14090)),
 ((96, 46), (42, 79), (12408, 11362)),
 ((65, 27), (12, 51), (8809, 6422)),
 ((11, 91), (66, 82), (5830, 9254)),
 ((57, 13), (16, 67), (5153, 5103)),
 ((11, 70), (83, 20), (14755, 16410)),
 ((51, 84), (66, 23), (5607, 2550)),
 ((14, 59), (78, 20), (14912, 3331)),
 ((56, 21), (25, 58), (4872, 6778)),
 ((75, 40), (22, 40), (7732, 5000)),
 ((15, 41), (82, 51), (5669, 14331)),
 ((98, 34), (75, 89), (10561, 9899)),
 ((12, 49), (43, 13), (14958, 12094)),
 ((18, 52), (48, 21), (12410, 2602)),
 ((77, 17), (16, 60), (7055, 2247)),
 ((13, 40), (43, 24), (4112, 9984)),
 ((37, 75), (68, 37), (8408, 7867)),
 ((13, 42), (70, 38), (7381, 302)),
 ((28, 56), (53, 11), (8944, 14348)),
 ((91, 43), (12, 47), (9012, 5209)),
 ((13, 78), (70, 46), (5613, 8620)),
 ((17, 42), (89, 55), (4722, 3422)),
 ((85, 48), (28, 73), (8605, 6289)),
 ((13, 38), (53, 36), (1219

In [25]:
def solver(problem):
    (a_x, a_y), (b_x, b_y), (prize_x, prize_y) = problem
    min_cost = None
    for a in range(101):
        d_x = prize_x - a_x * a
        d_y = prize_y - a_y * a
        if d_x < 0 or d_y < 0:
            break
        if d_x / b_x == d_y / b_y: # solution exists for a
            if d_x == b_x * (d_x // b_x): # solution is integer
                b = d_x // b_x
                if b <= 100:
                    cost = 3 * a + b
                    if not min_cost or cost < min_cost:
                        min_cost = cost
    return min_cost

In [26]:
total_cost = 0
for machine in data:
    if cost := solver(machine):
        total_cost += cost
total_cost

40369

Your puzzle answer was 40369.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
As you go to win the first prize, you discover that the claw is nowhere near where you expected it would be. Due to a unit conversion error in your measurements, the position of every prize is actually 10000000000000 higher on both the X and Y axis!

Add 10000000000000 to the X and Y position of every prize. After making this change, the example above would now look like this:

Button A: X+94, Y+34
Button B: X+22, Y+67
Prize: X=10000000008400, Y=10000000005400

Button A: X+26, Y+66
Button B: X+67, Y+21
Prize: X=10000000012748, Y=10000000012176

Button A: X+17, Y+86
Button B: X+84, Y+37
Prize: X=10000000007870, Y=10000000006450

Button A: X+69, Y+23
Button B: X+27, Y+71
Prize: X=10000000018641, Y=10000000010279
Now, it is only possible to win a prize on the second and fourth claw machines. Unfortunately, it will take many more than 100 presses to do so.

Using the corrected prize coordinates, figure out how to win as many prizes as possible. What is the fewest tokens you would have to spend to win all possible prizes?

In [28]:
conv_err = 10000000000000
with open('input13.txt') as file:
    data = []
    for line in file: # all inputs are positive
        if line.startswith('Button A'):
            btn_a = tuple(int(x) for x in re.findall(r'\d+', line)) 
        elif line.startswith('Button B'):
            btn_b = tuple(int(x) for x in re.findall(r'\d+', line))
        elif line.startswith('Prize'):
            prize = tuple(int(x) + conv_err for x in re.findall(r'\d+', line))
            data.append((btn_a, btn_b, prize))
data


[((40, 38), (21, 84), (10000000004245, 10000000005634)),
 ((19, 11), (37, 98), (10000000003246, 10000000006474)),
 ((12, 38), (44, 28), (10000000000812, 10000000012604)),
 ((29, 13), (39, 72), (10000000002243, 10000000014090)),
 ((96, 46), (42, 79), (10000000012408, 10000000011362)),
 ((65, 27), (12, 51), (10000000008809, 10000000006422)),
 ((11, 91), (66, 82), (10000000005830, 10000000009254)),
 ((57, 13), (16, 67), (10000000005153, 10000000005103)),
 ((11, 70), (83, 20), (10000000014755, 10000000016410)),
 ((51, 84), (66, 23), (10000000005607, 10000000002550)),
 ((14, 59), (78, 20), (10000000014912, 10000000003331)),
 ((56, 21), (25, 58), (10000000004872, 10000000006778)),
 ((75, 40), (22, 40), (10000000007732, 10000000005000)),
 ((15, 41), (82, 51), (10000000005669, 10000000014331)),
 ((98, 34), (75, 89), (10000000010561, 10000000009899)),
 ((12, 49), (43, 13), (10000000014958, 10000000012094)),
 ((18, 52), (48, 21), (10000000012410, 10000000002602)),
 ((77, 17), (16, 60), (10000000

In [ ]:
from math import isclose
# not needed for solution since no linearly dependent machines:
def find_ints(max_par, min_par, target):
    first_approx = target // max_par
    for x in range(first_approx, -1, -1):
        for y in range(0, target // min_par + 1):
            if isclose(x * max_par + y * min_par, target):
                return (x, y)
            elif x * max_par + y * min_par > target:
                break
    return (None, None)

In [73]:
total_cost = 0
for machine in data:
    (a_x, a_y), (b_x, b_y), (c_x, c_y) = machine

    # check for linear dependence:
    if isclose(a_x / a_y, b_x / b_y):
        print('yay')
        if a_x == a_y:
            print('ay')
        # no linearly dependent buttons, therefore this code is irrelevant:
        # if isclose(c_x / a_x, c_y / a_y): # solution exists:
        #     if b_x >= a_x / 3: # btn B cheaper per unit than btn A
        #         b, a = find_ints(b_x, a_x, c_x)
        #     else:
        #         a, b = find_ints(a_x, b_x, c_x)
        #     if a:
        #         cost = 3 * a + b
        #         total_cost += cost

    else:
        a = (b_x * c_y - b_y * c_x) / (a_y * b_x - a_x * b_y)
        b = (c_y - a * a_y) / b_y
        if a >= 0 and b >= 0 and isclose(a, round(a), abs_tol=0.0001, rel_tol=0) and isclose(b, round(b), abs_tol=0.0001, rel_tol=0):
            a = round(a)
            b = round(b)
            if c_x != a * a_x + b * b_x or c_y != a * a_y + b * b_y:
                print('Error!')
            cost = 3 * a + b
            total_cost += cost
total_cost

72587986598368

That's the right answer! You are one gold star closer to finding the Chief Historian.

